dot/types/grandpa.go

```diff
@@ -196,3 +196,62 @@ type GrandpaVote struct {
 func (v GrandpaVote) String() string {
 	return fmt.Sprintf("hash=%s number=%d", v.Hash, v.Number)
 }
+
+// GrandpaEquivocation is used to create a proof of equivocation
+type GrandpaEquivocation struct {
+	RoundNumber     uint64
+	ID              [32]byte
+	FirstVote       GrandpaVote
+	FirstSignature  [64]byte
+	SecondVote      GrandpaVote
+	SecondSignature [64]byte
+}
+
+// GrandpaEquivocationEnum is a wrapper object for GRANDPA equivocation proofs, useful for unifying prevote
+// / and precommit equivocations under a common type.
+type GrandpaEquivocationEnum scale.VaryingDataType
```

Comment:
```suggestion
type GrandpaEquivocationEnum scale.VaryingDataType
```
I think this should be named `GrandpaEquivocation` instead of on line 201.  It reflects what is in the `primitives/finality-grandpa` package in substrate.  You can make the struct on line 201 a private base struct and redefine the `PreVoteEquivocation` and `PrecommitEquivocation` definitions.

Reply:
Unless I am misunderstanding, this causes issues because the current GrandpaEquivocation struct on line 201 is used in other packages. What I think we can do is when we have the standalone grandpa package, they can both be GrandpaEquivocation. I do believe this is what substrate does

Reply:
I see.  There are two types named `Equivocation` in substrate.  One is in `sp_finality_grandpa::Equivocation` and one is `finality_grandpa::Equivocation`.  The `sp_finality_grandpa::Equivocation` type references `finality_grandpa::Equivocation`.  I think the types within the `types` package should hold `sp_finality_grandpa::Equivocation` and we should move the type on line 201 to the `grandpa` package which will be replaced by the standalone grandpa package in the future.